File: sdks/java/Wire.java
// Copyright 2026 J3nna Technologies, LLC
// SPDX-License-Identifier: Apache-2.0
//
// The J3nna Mesh wire layer for Java/the JVM: the canonical signing bytes every peer must reproduce, plus
// ed25519 sign/verify. This is the byte-for-byte contract — validated against the shared
// jip/conformance/vectors.json by ConformanceTest, so a Java peer is wire-compatible with the Go reference
// (and therefore every other SDK).
//
// Framing primitive: variable-length fields are 4-byte big-endian length-prefixed; integers are 8-byte
// big-endian; sets (capabilities, scopes) are sorted before framing.
//
// NB: this SDK is conceptually the `mesh` package, but the files are kept in the default package (no
// `package` declaration) so the conformance build `javac -cp gson.jar *.java && java -cp .:gson.jar
// ConformanceTest` works exactly as in jip/conformance/java — see README.

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.SecureRandom;
import java.security.Signature;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

public final class Wire {
    public static final String PROTOCOL = "JIP/0.1";
    public static final int PROTOCOL_MAJOR = 1;
    public static final String SIG_ALG = "ed25519";

    private static final SecureRandom RNG = new SecureRandom();

    private Wire() {}

    // --- framing primitives (identical to the conformance) ---

    static void field(ByteArrayOutputStream b, byte[] x) {
        try {
            b.write(ByteBuffer.allocate(4).putInt(x.length).array());
            b.write(x);
        } catch (IOException e) {
            throw new UncheckedIOException(e); // ByteArrayOutputStream never throws
        }
    }

    static void u64(ByteArrayOutputStream b, long v) {
        b.write(ByteBuffer.allocate(8).putLong(v).array(), 0, 8);
    }

    static void u32(ByteArrayOutputStream b, int v) {
        b.write(ByteBuffer.allocate(4).putInt(v).array(), 0, 4);
    }

    static byte[] U(String s) { return s.getBytes(StandardCharsets.UTF_8); }

    private static String alg(String a) { return (a == null || a.isEmpty()) ? SIG_ALG : a; }

    // --- signing-byte builders (mirror wire.py) ---

    public static byte[] presenceSigningBytes(String protocol, String alg, String id, byte[] publicKey,
                                              String endpoint, String mcpPath, List<String> capabilities,
                                              int protocolMajor, String grantId, long heartbeatUnix) {
        var b = new ByteArrayOutputStream();
        field(b, U(protocol));
        field(b, U(alg(alg)));
        field(b, U(id));
        field(b, publicKey);
        field(b, U(endpoint));
        field(b, U(mcpPath));
        var caps = new ArrayList<>(capabilities);
        Collections.sort(caps);
        u32(b, caps.size());
        for (var c : caps) field(b, U(c));
        u32(b, protocolMajor);
        field(b, U(grantId == null ? "" : grantId));
        u64(b, heartbeatUnix);
        return b.toByteArray();
    }

    public static byte[] grantSigningBytes(String alg, String id, String subject, byte[] publicKey, long tier,
                                           List<String> scopes, long issuedAt, long notAfter, String principal) {
        var b = new ByteArrayOutputStream();
        field(b, U("J3nna-mesh-grant/1"));
        field(b, U(alg(alg)));
        field(b, U(id));
        field(b, U(subject));
        field(b, publicKey);
        u64(b, tier);
        var sc = scopes == null ? new ArrayList<String>() : new ArrayList<>(scopes);
        Collections.sort(sc);
        field(b, U(String.join("\0", sc)));
        u64(b, issuedAt);
        u64(b, notAfter);
        if (principal != null && !principal.isEmpty()) { // signature-covered only when present
            field(b, U("J3nna-mesh-principal/1"));
            field(b, U(principal));
        }
        return b.toByteArray();
    }

    public static byte[] callproofSigningBytes(String alg, String nodeId, String tool, byte[] argsHash,
                                               long unixMilli) {
        var b = new ByteArrayOutputStream();
        field(b, U("JIP-call/0.2"));
        field(b, U(alg(alg)));
        field(b, U(nodeId));
        field(b, U(tool));
        field(b, argsHash);
        u64(b, unixMilli);
        return b.toByteArray();
    }

    /** Field-framed, signed by the NODE key to prove possession of the pinned identity. */
    public static byte[] renewSigningBytes(String alg, String grantId, String subject, byte[] publicKey,
                                           long issuedAt) {
        var b = new ByteArrayOutputStream();
        field(b, U("J3nna-mesh-renew/1"));
        field(b, U(alg(alg)));
        field(b, U(grantId));
        field(b, U(subject));
        field(b, publicKey);
        u64(b, issuedAt);
        return b.toByteArray();
    }

    /**
     * NOT field-framed: pipe/comma ASCII with a trailing comma after EVERY id; ids sorted ascending.
     *   J3nna-mesh-crl/1|&lt;alg&gt;|&lt;issued_at&gt;|&lt;id1&gt;,&lt;id2&gt;,...,
     */
    public static byte[] crlSigningBytes(String alg, long issuedAt, List<String> revokedIds) {
        StringBuilder head = new StringBuilder("J3nna-mesh-crl/1|")
                .append(alg(alg)).append("|").append(issuedAt).append("|");
        var ids = new ArrayList<>(revokedIds);
        Collections.sort(ids);
        for (String rid : ids) head.append(rid).append(",");
        return U(head.toString());
    }

    // --- canonical args JSON / hash (the one cross-language gotcha) ---

    private static String jsonString(String s) {
        return "\"" + s.replace("\\", "\\\\").replace("\"", "\\\"") + "\"";
    }

    /**
     * Reproduce Go's json.Marshal of the arguments map: keys sorted, compact, and {@code < > &} escaped as
     * {@code < > &}. Values are rendered by gson's compact serialization (whole numbers as
     * integers, since gson keeps Number formatting). Mirror of wire.py's canonical_args_json.
     */
    public static String canonicalArgsJson(Map<String, Object> args) {
        var sorted = new TreeMap<>(args);
        var sb = new StringBuilder("{");
        boolean first = true;
        for (var e : sorted.entrySet()) {
            if (!first) sb.append(",");
            first = false;
            sb.append(jsonString(e.getKey())).append(":").append(renderValue(e.getValue()));
        }
        sb.append("}");
        return sb.toString().replace("<", "\\u003c").replace(">", "\\u003e").replace("&", "\\u0026");
    }

    /** Canonicalize a gson JsonObject directly (used by the conformance harness). */
    public static String canonicalArgsJson(com.google.gson.JsonObject args) {
        var keys = new ArrayList<>(args.keySet());
        Collections.sort(keys);
        var sb = new StringBuilder("{");
        for (int k = 0; k < keys.size(); k++) {
            if (k > 0) sb.append(",");
            var key = keys.get(k);
            var v = args.get(key);
            var vs = (v.isJsonPrimitive() && v.getAsJsonPrimitive().isString())
                    ? jsonString(v.getAsString()) : v.toString();
            sb.append(jsonString(key)).append(":").append(vs);
        }
        sb.append("}");
        return sb.toString().replace("<", "\\u003c").replace(">", "\\u003e").replace("&", "\\u0026");
    }

    private static String renderValue(Object v) {
        if (v == null) return "null";
        if (v instanceof String s) return jsonString(s);
        if (v instanceof Boolean b) return b.toString();
        if (v instanceof Integer || v instanceof Long || v instanceof Short || v instanceof Byte) {
            return v.toString();
        }
        if (v instanceof Number n) {
            // Render whole-valued floats/doubles without a decimal point, matching Go integers.
            double d = n.doubleValue();
            if (d == Math.rint(d) && !Double.isInfinite(d)) return Long.toString((long) d);
            return n.toString();
        }
        // Fall back to gson for any nested structure.
        return new com.google.gson.Gson().toJson(v);
    }

    public static byte[] argsHash(Map<String, Object> args) {
        try {
            return MessageDigest.getInstance("SHA-256").digest(U(canonicalArgsJson(args)));
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    // --- ed25519 sign / verify ---

    public static byte[] sign(PrivateKey priv, byte[] msg) {
        try {
            var s = Signature.getInstance("Ed25519");
            s.initSign(priv);
            s.update(msg);
            return s.sign();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public static boolean verify(PublicKey pub, byte[] sig, byte[] msg) {
        try {
            var s = Signature.getInstance("Ed25519");
            s.initVerify(pub);
            s.update(msg);
            return s.verify(sig);
        } catch (Exception e) {
            return false;
        }
    }

    /** Wrap a raw 32-byte ed25519 public key in X.509 SPKI and build a PublicKey (the conformance approach). */
    public static PublicKey ed25519Pub(byte[] raw) {
        try {
            byte[] prefix = unhex("302a300506032b6570032100");
            byte[] der = new byte[prefix.length + raw.length];
            System.arraycopy(prefix, 0, der, 0, prefix.length);
            System.arraycopy(raw, 0, der, prefix.length, raw.length);
            return java.security.KeyFactory.getInstance("Ed25519")
                    .generatePublic(new java.security.spec.X509EncodedKeySpec(der));
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    // --- trace ids ---

    public static String newSpanId() { return randomHex(8); }

    /** A fresh W3C traceparent (version 00, sampled). */
    public static String newTraceparent() {
        return "00-" + randomHex(16) + "-" + randomHex(8) + "-01";
    }

    private static String randomHex(int nBytes) {
        byte[] b = new byte[nBytes];
        RNG.nextBytes(b);
        return enhex(b);
    }

    // --- hex helpers (shared) ---

    public static byte[] unhex(String s) {
        byte[] o = new byte[s.length() / 2];
        for (int k = 0; k < o.length; k++) {
            o[k] = (byte) Integer.parseInt(s.substring(2 * k, 2 * k + 2), 16);
        }
        return o;
    }

    public static String enhex(byte[] b) {
        var sb = new StringBuilder();
        for (byte x : b) sb.append(String.format("%02x", x));
        return sb.toString();
    }
}

File: sdks/python/src/j3nna_mesh/wire.py
# Copyright 2026 J3nna Technologies, LLC
# SPDX-License-Identifier: Apache-2.0
"""The J3nna Mesh wire layer for Python: the canonical signing bytes every peer must reproduce, plus
ed25519 sign/verify. This is the byte-for-byte contract — it is validated against the shared
jip/conformance/vectors.json by tests/test_conformance.py, so a Python peer is wire-compatible with the Go
reference (and therefore every other SDK).

Framing primitive: variable-length fields are 4-byte big-endian length-prefixed; integers are 8-byte
big-endian; sets (capabilities, scopes) are sorted before framing. Two structures (CRL, renew) use their
own framings, documented inline.
"""

import hashlib
import json
import secrets
import struct

from cryptography.hazmat.primitives.asymmetric import ed25519

PROTOCOL = "JIP/0.1"
PROTOCOL_MAJOR = 1
SIG_ALG = "ed25519"


def _field(buf: bytearray, x: bytes) -> None:
    buf += struct.pack(">I", len(x))
    buf += x


def _u64(buf: bytearray, v: int) -> None:
    buf += struct.pack(">Q", v & 0xFFFFFFFFFFFFFFFF)


def _u32(buf: bytearray, v: int) -> None:
    buf += struct.pack(">I", v & 0xFFFFFFFF)


def _alg(a: str) -> bytes:
    return (a or SIG_ALG).encode()


def presence_signing_bytes(*, protocol, alg, id, public_key, endpoint, mcp_path, capabilities,
                           protocol_major, grant_id, heartbeat_unix) -> bytes:
    b = bytearray()
    _field(b, protocol.encode())
    _field(b, _alg(alg))
    _field(b, id.encode())
    _field(b, public_key)
    _field(b, endpoint.encode())
    _field(b, mcp_path.encode())
    caps = sorted(capabilities)
    _u32(b, len(caps))
    for c in caps:
        _field(b, c.encode())
    _u32(b, protocol_major)
    _field(b, (grant_id or "").encode())
    _u64(b, heartbeat_unix)
    return bytes(b)


def grant_signing_bytes(*, alg, id, subject, public_key, tier, scopes, issued_at, not_after,
                        principal="") -> bytes:
    b = bytearray()
    _field(b, b"J3nna-mesh-grant/1")
    _field(b, _alg(alg))
    _field(b, id.encode())
    _field(b, subject.encode())
    _field(b, public_key)
    _u64(b, tier)
    _field(b, b"\x00".join(s.encode() for s in sorted(scopes or [])))
    _u64(b, issued_at)
    _u64(b, not_after)
    if principal:  # signature-covered only when present, so legacy grants verify byte-identically
        _field(b, b"J3nna-mesh-principal/1")
        _field(b, principal.encode())
    return bytes(b)


def callproof_signing_bytes(*, alg, node_id, tool, args_hash, unix_milli) -> bytes:
    b = bytearray()
    _field(b, b"JIP-call/0.2")
    _field(b, _alg(alg))
    _field(b, node_id.encode())
    _field(b, tool.encode())
    _field(b, args_hash)
    _u64(b, unix_milli)
    return bytes(b)


def renew_signing_bytes(*, alg, grant_id, subject, public_key, issued_at) -> bytes:
    # Field-framed, signed by the NODE key to prove possession of the pinned identity.
    b = bytearray()
    _field(b, b"J3nna-mesh-renew/1")
    _field(b, _alg(alg))
    _field(b, grant_id.encode())
    _field(b, subject.encode())
    _field(b, public_key)
    _u64(b, issued_at)
    return bytes(b)


def crl_signing_bytes(*, alg, issued_at, revoked_ids) -> bytes:
    # NOT field-framed: pipe/comma ASCII with a trailing comma after EVERY id; ids sorted ascending.
    #   J3nna-mesh-crl/1|<alg>|<issued_at>|<id1>,<id2>,...,
    head = f"J3nna-mesh-crl/1|{alg or SIG_ALG}|{issued_at}|"
    body = "".join(f"{rid}," for rid in sorted(revoked_ids))
    return (head + body).encode()


def canonical_args_json(args: dict) -> bytes:
    """Reproduce Go's json.Marshal of the arguments map: keys sorted, compact, and <, >, & escaped as
    \\u003c \\u003e \\u0026. This is the one place JSON canonicalization must match byte-for-byte."""
    s = json.dumps(args, sort_keys=True, separators=(",", ":"), ensure_ascii=False)
    return s.replace("<", "\\u003c").replace(">", "\\u003e").replace("&", "\\u0026").encode()


def args_hash(args: dict) -> bytes:
    return hashlib.sha256(canonical_args_json(args)).digest()


def sign(seed32: bytes, msg: bytes) -> bytes:
    return ed25519.Ed25519PrivateKey.from_private_bytes(seed32).sign(msg)


def verify(public_key32: bytes, sig: bytes, msg: bytes) -> bool:
    try:
        ed25519.Ed25519PublicKey.from_public_bytes(public_key32).verify(sig, msg)
        return True
    except Exception:
        return False


def new_span_id() -> str:
    """A fresh 64-bit span id (16 hex chars)."""
    return secrets.token_hex(8)


def new_traceparent() -> str:
    """A fresh W3C `traceparent` (version 00, sampled). Attach one across a logical operation's calls so a
    telemetry backend stitches them into a single trace."""
    return "00-" + secrets.token_hex(16) + "-" + secrets.token_hex(8) + "-01"
